3rdparty/mshadow/mshadow/base.h

```diff
@@ -606,6 +606,64 @@ struct divto {
   typedef op::div OPType;
 };
 }  // namespace sv
+
+#ifndef __CUDA_ARCH__
+using std::isnan;
+using std::isinf;
+#endif
+
+/*! \brief
+ *  determines if the given floating point
+ *  number is not a number */
+namespace isnan_typed {
+  template<typename DType>
+  MSHADOW_XINLINE bool IsNan(volatile DType val) {
+    return false;
+  }
+  template<>
+  MSHADOW_XINLINE bool IsNan(volatile float val) {
+    return isnan(val);
+  }
+  template<>
+  MSHADOW_XINLINE bool IsNan(volatile double val) {
+    return isnan(val);
+  }
+  template<>
+  MSHADOW_XINLINE bool IsNan(volatile long double val) {
+    return isnan(val);
+  }
+  template<>
+  MSHADOW_XINLINE bool IsNan(volatile mshadow::half::half_t val) {
+    return (val.half_ & 0x7fff) > 0x7c00;
```

Comment:
Can you turn these magic values into constants with documentation? While I get 0x7ffff, 0x7c00 for example, looks quite arbitrary.

Reply:
Hi @marcoabreu , I add two constants `MSHADOW_HALF_SIGN_BIT` and `MSHADOW_HALF_EXPONENT_BITS` in [3rdparty/mshadow/mshadow/half.h](https://github.com/apache/incubator-mxnet/pull/16234/commits/9183cb3a2284a7d920bb720a19c38dc49d66fdea#diff-0d54cfcb67beaf0744f1efe71b45bfdfR352), and replace these two magic values.